### ISSUE TITLE
Support for ValueTask

### ISSUE THREAD

**Issue by benmccallum** [shouldly/shouldly#753]
Sorry if my knowledge in this space makes this all invalid, but I've recently needed to start writing tests against `ValueTask`-returning methods and was surprised that I couldn't do:
```csharp
Should.ThrowAsync(AsyncFunc());
```

In a few places I was able to tack on the end `.AsTask()` to make it fit the existing `Task` overloads:
```csharp
Should.ThrowAsync(AsyncFunc().AsTask());
```

But this seems to fall over when the method is only doing sync things as I believe the `ValueTask` optimization is such that when `.AsTask()` is called, it knows it's sync and immediately fires the method, causing the exception to be thrown in a way that's not "handled" by Shouldly, e.g. it'll just throw and fail rather than throw but be handled by the assertion.

I'm not sure what the best path forward here is, but just wanted to start the discussion as I didn't see any previous issues referring to `ValueTask` support.

Here's a complete example:
```csharp
using System;
using System.Threading.Tasks;
using Shouldly;
using Xunit;

public class Repro
{
    [Fact]
    public async Task DoesntWork()
    {
        await Should.ThrowAsync<Exception>(SyncFunc().AsTask());
    }

    [Fact]
    public async Task Works()
    {
        await Should.ThrowAsync<Exception>(AsyncFunc().AsTask());
    }

    private ValueTask SyncFunc()
    {
        throw new Exception();
        return default;
    }

    private async ValueTask AsyncFunc()
    {
        throw new Exception();
        await Task.FromResult(0);
    }
}
```

**Comment by benmccallum** [comment]
Not sure how to title this without sounding patronizing whilst acknowledging the fact there is support with .AsTask() but I've seemingly to find a loophole. Hope you get the gist. Thanks :)

**Comment by benmccallum** [comment]
I've just realised wrapping it in another func is a decent workaround:
```csharp
await Should.ThrowAsync<Exception>(() => SyncFunc().AsTask());
```

**Comment by jnm2** [comment]
I think our support for ValueTask would amount to overloads doing the same workaround. We need to be cautious of maintaining an explosion of overloads, but if we would consider paying that cost to help reduce repetition on the consumption side if it is both painful enough and common enough.

> But this seems to fall over when the method is only doing sync things as I believe the ValueTask optimization is such that when .AsTask() is called, it knows it's sync and immediately fires the method, causing the exception to be thrown in a way that's not "handled" by Shouldly, e.g. it'll just throw and fail rather than throw but be handled by the assertion.

This should not be the case. Both Task and ValueTask sometimes throw unwrapped exceptions and sometimes return ValueTask/Task.FromException. Throwing directly is best for exceptions that say "the caller should have known better," like ArgumentException, because these should not be caught unless you're dealing with a poorly designed API that forces you to catch them. Returning a Task or ValueTask instance wrapping an exception is necessary in order to make it possible to handle the exception property if the task is created at a different point in the code than where it is awaited.

So, `Should.Throw<Exception>(TaskReturningMethod())` calls `TaskReturningMethod()` before Shouldly has a chance to get involved, and therefore you'd see the exception being thrown outside Shouldly's control whether or not the method returns ValueTask. In comparison, `Should.Throw<Exception>(TaskReturningMethod)` calls a different Shouldly overload where `Should.Throw` gets called first and then it can decide when to invoke the passed callback.

tl;dr Task- and ValueTask-returning methods *should* behave consistently to each other in this regard, and it's not related to Shouldly's support or lack of support for ValueTask.

**Comment by benmccallum** [comment]
Thanks for the detailed response @jnm2. 

The distinction between the following is something I didn't realise and will be aware of next time.
* `Should.Throw<Exception>(TaskReturningMethod())` vs 
* `Should.Throw<Exception>(TaskReturningMethod)` / `Should.Throw<Exception>(() => TaskReturningMethod())`

By no means feel obliged to support new overloads if there's not enough support from others. I think we could document this somewhere though to demonstrate how to handle `ValueTask` and touch on the above distinctions.

**Comment by jnm2** [comment]
Another thing we could consider is obsoleting the `Should.Throw<T>(Task)` overloads and requiring `Should.Throw<T>(Func<Task>)` only. That would make the syntax `Should.Throw<Exception>(TaskReturningMethod())` stop working, requiring one of the other two syntaxes to be used, and possibly avoid this pitfall. Having fewer ways to do the same thing speeds up developers.

**Comment by benmccallum** [comment]
> obsoleting the Should.Throw<T>(Task)

I'd be in favour of that, makes sense.

**Comment by jnm2** [comment]
Cool. That's now tracked at https://github.com/shouldly/shouldly/issues/789.

**Comment by Shane32** [comment]
In the past when I've encountered situations where I'm considering overloads for both Task and ValueTask, I've found that the compiler cannot determine which overload to use with an inline delegate, which makes it impractical in use cases like this.

```cs
// so, this wouldn't compile if there were two overloads, one Task-based and one ValueTask-based
await Should.ThrowAsync<InvalidOperationException>(async () => await HelloAsync("world"));

// this would...
Func<Task> func = async () => await HelloAsync("world");
await Should.ThrowAsync<InvalidOperationException>(func);

// and this would...
Func<ValueTask> func = async () => await HelloAsync("world");
await Should.ThrowAsync<InvalidOperationException>(func);

// and this would...
await Should.ThrowAsync<InvalidOperationException>((Func<Task>)(async () => await HelloAsync("world")));
```

This restriction does not exist for extension methods, but we can't write an extension method for a static class.

I wouldn't add ValueTask overloads due to this reason.
